src/WindowsUtils/Debugger.cpp

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2022 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "WindowsUtils/Debugger.h"
+
+#include "OrbitBase/GetLastError.h"
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/StringConversion.h"
+#include "OrbitBase/ThreadUtils.h"
+#include "WindowsUtils/CreateProcess.h"
+#include "WindowsUtils/SafeHandle.h"
+
+namespace orbit_windows_utils {
+
+Debugger::Debugger() : process_info_or_error_(ErrorMessage()) {}
+Debugger::~Debugger() { Wait(); }
+
+ErrorMessageOr<ProcessInfo> Debugger::Start(const std::filesystem::path& executable,
+                                            const std::filesystem::path& working_directory,
+                                            std::string_view arguments) {
+  // Launch process and start debugging loop on the same separate thread.
+  thread_ = std::thread(&Debugger::DebuggerThread, this, executable, working_directory,
+                        std::string(arguments));
+
+  // We use "create_process_promise_" to wait until the result of "CreateProcessToDebug", which is
+  // stored in "process_info_or_error_", is available from the  debugger thread.
+  create_process_promise_.GetFuture().Wait();
+  return std::move(process_info_or_error_);
```

Comment:
You should bring `process_info_or_error_` into a knonw valid state again by first moving the value into a local variable, then clearing `process_info_or_error_`, then returning the local variable. Due to guaranted copy elision this won't imply any additional copy or move.

Reply:
not applicable after changes.